src/sdk/cpp/lib/src/protocol_gateway.cpp

```diff
@@ -0,0 +1,215 @@
+/*****************************************************************************
+ * Copyright (c) 2021 Bosch.IO GmbH
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ ****************************************************************************/
+#include "dlfcn.h"
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+
+#include "logging.hpp"
+#include "protocol_gateway.hpp"
+
+namespace iotea {
+namespace core {
+
+///////////////////
+// PubSubOptions //
+///////////////////
+PubSubOptions::PubSubOptions(bool platform_proto_only, const std::string& adapter_id)
+    : platform_proto_only_{platform_proto_only}
+    , adapter_id_{adapter_id} {
+
+}
+
+bool PubSubOptions::IsPlatformProtoOnly() const {
+    return platform_proto_only_;
+}
+
+std::string PubSubOptions::GetAdapterId() const {
+    return adapter_id_;
+}
+
+////////////////////
+// PublishOptions //
+////////////////////
+PublishOptions::PublishOptions(bool platform_proto_only, const std::string& adapter_id)
+    : PubSubOptions{platform_proto_only, adapter_id}
+    , retain_{false}
+    , stash_{true} {}
+
+bool PublishOptions::Retain() const {
+    return retain_;
+}
+
+bool PublishOptions::Stash() const {
+    return stash_;
+}
+
+//////////////////////
+// SubscribeOptions //
+//////////////////////
+SubscribeOptions::SubscribeOptions(bool platform_proto_only, const std::string& adapter_id)
+    : PubSubOptions{platform_proto_only, adapter_id} {}
+
+/////////////
+// Adapter //
+/////////////
+Adapter::Adapter(const std::string& name, bool is_platform_proto)
+    : name_{name}
+    , is_platform_proto_{is_platform_proto} {}
+
+std::string Adapter::GetName() const {
+    return name_;
+}
+
+bool Adapter::IsPlatformProto() const {
+    return is_platform_proto_;
+}
+
+//////////////////////////////
+// ProtocolGatewayException //
+//////////////////////////////
+ProtocolGatewayException::ProtocolGatewayException(const std::string& msg, const Code& code)
+    : msg_{msg}
+    , code_{code} {}
+
+const char* ProtocolGatewayException::what() const noexcept {
+    return msg_.c_str();
+}
+
+ProtocolGatewayException::Code ProtocolGatewayException::GetCode() const noexcept {
+    return code_;
+}
+
+/////////////////////
+// ProtocolGateway //
+/////////////////////
+const PublishOptions ProtocolGateway::DefaultPublishOptions{false, ""};
+
+const SubscribeOptions ProtocolGateway::DefaultSubscribeOptions{false, ""};
+
+void ProtocolGateway::ValidateConfig(const json& config, bool platform_proto_only) {
```

Comment:
Does it make sense to add a flag to activate / deactivate the validation (e.g. activated for debug, deactivated for release mode). Can be done in an upcoming PR :)

Reply:
Maybe, this is cheap check and modelled after what the Python SDK is doing.